### PR TITLE
Added Wazero dependent Wasm SDK

### PR DESCRIPTION
**Pull request by Kaijlo**
Added a [Wazero](https://github.com/tetratelabs/wazero) dependent Wasm SDK that is almost interchangeable with the Wasmtime dependent SDK.
Currently supports Go 1.17 and 1.18 as Wazero doesn't support Go 1.16

### CODE REVIEW

File: internal/wasm/wazero_sdk/internal/wazero/module.go [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+	m.env.Memory().Write(m.ctx, uint32(addr), data)
+
+	return uint32(addr)
+}
+
+//reads a null terminated string starting at the given address in the shared memory buffer
+func (m *Module) readStr(loc uint32) string {
+	bytes := []byte{}
+	var index uint32
+	for ok := true; ok; {
```
Comment (srenatus) [inline internal/wasm/wazero_sdk/internal/wazero/module.go:280]: 🤔 Hmm I feel like we could use `readUntil(loc, 0b0)` here -- can't we?

File: internal/wasm/wazero_sdk/internal/wazero/VM.go [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+			return nil, err
+		}
+		heapPtr += inputLen
+		metrics.Timer("wasm_vm_eval_prepare_input").Stop()
+	}
+	// Setting the ctx here ensures that it'll be available to builtins that
+	// make use of it (e.g. `http.send`); and it will spawn a go routine
+	// cancelling the builtins that use topdown.Cancel, when the context is
+	// cancelled.
+	i.module.Reset(ctx, seed, ns, iqbCache, ph, capabilities)
```
Comment (srenatus) [inline internal/wasm/wazero_sdk/internal/wazero/VM.go:401]: I'm afraid it's been a while since I've worked on the wasmtime-go integration -- so please bear with me -- Is it the right thing to do to reset the rng seed on every Eval?

File: internal/wasm/wazero_sdk/internal/wazero/builtins.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	builtinNameMap := parseJSONString(builtinStr)
+	builtinIDMap, err := getFuncs(builtinNameMap)
+	if err != nil {
+		panic(err)
+	}
+	return builtinIDMap
+}
+
+// json string parser
+func parseJSONString(str string) map[string]int32 {
```
Comment (srenatus) [inline internal/wasm/wazero_sdk/internal/wazero/builtins.go:30]: What's the reason not to use sdlib's encoding/json here?

**Review by srenatus — COMMENTED** [review]
Had a first read through. This is cool stuff, thanks for working on it :+1:

There are some large amounts of copied-over code here; which in itself is not a problem if refactoring is too cumbersome here -- but we should perhaps leave a few notes re: what was copied from where, and why.

Haven't had a chance to pull this down and play with it yet, but it the code looks mostly in a good shape. 🥳

File: Makefile [hunk 4]
```diff
@@ -144,6 +144,14 @@ perf-noisy: generate
...
 	$(GO) test $(GO_TAGS),slow,wasm_sdk_e2e $(GO_TEST_TIMEOUT) -v ./internal/wasm/sdk/test/e2e
 
+.PHONY: wasm-wazero-sdk-e2e-test
+wasm-wazero-sdk-e2e-test: generate
+	$(GO) test $(GO_TAGS),slow,wasm_sdk_e2e $(GO_TEST_TIMEOUT) -v ./internal/wasm/wazero_sdk/test/e2e
+
+.PHONY: wasm-wazero-sdk-e2e-test
+wasm-wazero-sdk-e2e-test: generate
+	$(GO) test $(GO_TAGS),slow,wasm_sdk_e2e $(GO_TEST_TIMEOUT) -v ./internal/wasm/wazero_sdk/test/e2e
+
```
[the remark this answers is not in the document]
Comment (Kaijlo) [reply Makefile:154]: The path is in the wazero_sdk instead of the sdk folder, the exceptions.yaml file differs from the wasmtime sdk

[the remark this answers is not in the document]
Comment (srenatus) [reply Makefile:154]: https://github.com/open-policy-agent/opa/pull/4821/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R147-R154 These two blocks still look the same to me 😅

File: internal/wasm/wazero_sdk/examples/basic/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+Basic Wasm Module Evaluation Example
```
[the remark this answers is not in the document]
Comment (Kaijlo) [reply internal/wasm/wazero_sdk/examples/basic/README.md:1]: I feel like they aren't necessary as it is used in exactly the same way as the wasmtime SDK

[the remark this answers is not in the document]
Comment (srenatus) [reply internal/wasm/wazero_sdk/examples/basic/README.md:1]: :+1: then let's drop them, please

File: internal/wasm/wazero_sdk/internal/wazero/VM.go (hunk 2 above)
Comment (Kaijlo) [reply internal/wasm/wazero_sdk/internal/wazero/VM.go:401]: It is what the wasmtime sdk does, although I have to assume that seed is passed as nil and is thus set to ns in unix timestamp format

File: internal/wasm/wazero_sdk/internal/wazero/VM.go [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+	vm.heapPtrSet = vm.module.heapPtrSet
+	vm.jsonDump = vm.module.jsonDump
+	vm.jsonParse = vm.module.jsonParse
+	vm.valueDump = vm.module.valueDump
+	vm.valueParse = vm.module.valueParse
+	vm.malloc = vm.module.malloc
+	vm.free = vm.module.free
+	vm.valueAddPath = vm.module.valueAddPath
+	vm.valueRemovePath = vm.module.valueRemovePath
+	err := vm.setData(vm.ctx, opts, "newVM")
```
[the remark this answers is not in the document]
Comment (Kaijlo) [reply internal/wasm/wazero_sdk/internal/wazero/VM.go:89]: I forgot to remove it after debugging, it was used for pinpointing an error

File: internal/wasm/wazero_sdk/internal/wazero/module.go [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+	seed                   io.Reader
+	ns                     time.Time
+	iqbCache               cache.InterQueryCache
+	ph                     print.Hook
+	capabilities           *ast.Capabilities
+}
+
+// Env is a wasm module that holds the shared memory buffer and the builtin bindings
+func (m *Module) newEnv(opts moduleOpts, r wazero.Runtime) (api.Module, error) {
+	if opts.maxMemSize == (moduleOpts{}).maxMemSize {
```
[the remark this answers is not in the document]
Comment (Kaijlo) [reply internal/wasm/wazero_sdk/internal/wazero/module.go]: Yea, that is much cleaner. I probably hadn't yet realized that the default value was 0 when I wrote this.

[the remark this answers is not in the document]
Comment (srenatus) [reply internal/wasm/wazero_sdk/internal/wazero/module.go]: Yeah you can't have default values there aren't just what the type's zero value is anyways 🤷

File: docs/content/management-bundles.md [hunk 8]
```diff
@@ -711,34 +711,6 @@ bundles:
 
 **NOTE:** the S3 `url` is the bucket's regional endpoint.
 
-##### Credential Provider Chaining
```
Comment (srenatus) [inline docs/content/management-bundles.md:714]: Looks like another rebase/merge mishap. LMK if you could use some assistance, git can be hard. 😅 But it shouldn't be what's eating your time.

File: go.mod [hunk 9]
```diff
@@ -29,7 +29,7 @@ require (
 	github.com/sirupsen/logrus v1.8.1
 	github.com/spf13/cobra v1.5.0
 	github.com/spf13/pflag v1.0.5
-	github.com/stretchr/testify v1.7.4
+	github.com/stretchr/testify v1.7.5
```
Comment (srenatus) [inline go.mod:32]: Looks unrelated; or also a rebase thing.

File: internal/wasm/wazero_sdk/internal/wazero/builtins.go [hunk 10]
```diff
@@ -17,50 +17,16 @@ func newBuiltinTable(mod Module) map[int32]topdown.BuiltinFunc {
 	if err != nil {
 		panic(err)
 	}
-	builtinStr := mod.readStr(uint32(builtinsJSON))
-	builtinNameMap := parseJSONString(builtinStr)
+	builtinStr := mod.readStr(builtinsJSON)
+	var builtinNameMap map[string]int32
+	json.Unmarshal([]byte(builtinStr), &builtinNameMap)
```
Comment (srenatus) [inline internal/wasm/wazero_sdk/internal/wazero/builtins.go:22]: Isn't there an error return here?

**Review by srenatus — COMMENTED** [review]
Looks like some wires got crossed rebasing this? But I've noted many good improvements :+1: thanks for bearing with me.

File: go.mod (hunk 9 above)
Comment (Kaijlo) [reply go.mod:32]: rebase thing

File: internal/wasm/wazero_sdk/internal/wazero/builtins.go (hunk 10 above)
Comment (Kaijlo) [reply internal/wasm/wazero_sdk/internal/wazero/builtins.go:22]: yea, I forgot it had any return values

File: plugins/rest/rest_auth.go [hunk 11]
```diff
@@ -520,65 +519,21 @@ type awsSigningAuthPlugin struct {
 	logger logging.Logger
 }
 
-type awsCredentialServiceChain struct {
```
[the remark this answers is not in the document]
Comment (Kaijlo) [reply plugins/rest/rest_auth.go:523]: that might have happened during rebasing

[the remark this answers is not in the document]
Comment (Kaijlo) [reply plugins/rest/rest_auth.go]: Should now be fixed

File: internal/wasm/wazero_sdk/internal/wazero/builtins.go (hunk 10 above)
Comment (Kaijlo) [reply internal/wasm/wazero_sdk/internal/wazero/builtins.go]: Added the error checks

File: docs/content/management-bundles.md [hunk 12]
```diff
@@ -112,7 +112,7 @@ bundle has not changed since the last update.
 With the periodic bundle downloading (ie. `short polling`) technique, OPA sends regular requests to the remote HTTP
```
Comment (srenatus) [inline docs/content/management-bundles.md:112]: Can you remove these changes, please? it's unrelated, looks like some linter made a pass over the file 😅

File: test/cases/testdata/aggregates/test-aggregates-0001.yaml [hunk 13]
```diff
@@ -5,6 +5,83 @@ cases:
     - 2
     - 3
     - 4
```
Comment (srenatus) [inline test/cases/testdata/aggregates/test-aggregates-0001.yaml:7]: 😎 Let's not bring these back please 😉

**Review by srenatus — COMMENTED** [review]
Still some git hurdles, it seems. Also go.sum and go.mod seem to have diverged, could you please rebase this again? 😃

**Comment by Kaijlo** [comment]
the differences in go.mod and go.sum seems to just be the wazero package

**Comment by srenatus** [comment]
ℹ️ It might take a little longer, but I'm adopting this PR.

**Comment by srenatus** [comment]
OK, I've got this. Closing the PR for now because our open PRs page is a mess 🧹
